fedimint-core/src/db/notifications.rs

```diff
@@ -0,0 +1,236 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+use anyhow::Context;
+use bitvec::vec::BitVec;
+use fedimint_core::{apply, async_trait_maybe_send};
+use tokio::sync::futures::Notified;
+use tokio::sync::Notify;
+
+use super::{ISingleUseDatabaseTransaction, PrefixStream, Result};
+
+/// Number of buckets used for `Notifications`.
+const NOTIFY_BUCKETS: usize = 32;
+
+/// The state of Notification.
+///
+/// This stores `NOTIFY_BUCKETS` number of `Notifies`.
+/// Each key is assigned a bucket based on its hash value.
+/// This will cause some false positives.
+#[derive(Debug)]
+pub struct Notifications {
+    buckets: Vec<Notify>,
+}
+
+impl Default for Notifications {
+    fn default() -> Self {
+        Self {
+            buckets: (0..NOTIFY_BUCKETS).map(|_| Notify::new()).collect(),
+        }
+    }
+}
+
+impl Notifications {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// This registers for notification when called.
+    ///
+    /// Then waits for the notification when .awaited.
+    ///
+    /// NOTE: This may some false positives.
+    pub fn register<K>(&self, key: &K) -> Notified
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.buckets.len() as u64)) as usize;
+        self.buckets[slot_idx].notified()
+    }
+
+    /// Notify a key.
+    ///
+    /// All the waiters for this keys will be notified.
+    pub async fn notify<K>(&self, key: &K)
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.buckets.len() as u64)) as usize;
+        self.buckets[slot_idx].notify_waiters();
+    }
+
+    /// Notifies the waiters about the notifications recorded in NotifyQueue.
+    pub fn submit_queue(&self, queue: NotifyQueue) {
+        for bucket in queue.buckets.iter_ones() {
+            self.buckets[bucket].notify_waiters();
+        }
+    }
+}
+
+/// Save notifications to be sent after transaction is complete.
+#[derive(Debug)]
+pub struct NotifyQueue {
+    buckets: BitVec,
+}
+
+impl Default for NotifyQueue {
+    fn default() -> Self {
+        Self {
+            buckets: BitVec::repeat(false, NOTIFY_BUCKETS),
+        }
+    }
+}
+
+impl NotifyQueue {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add<K>(&mut self, key: &K)
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.buckets.len() as u64)) as usize;
+        self.buckets.set(slot_idx, true);
+    }
+}
+
+/// Wrapper to implements `add_notification_key`
+pub struct NotifyingTransactionWrapper<'a> {
```

Comment:
Not sure if I like "Wrapper". Maybe `NotifyingDatabaseTransaction`? That could be clear